libyul/optimiser/SimplificationRules.cpp

```diff
@@ -171,12 +171,26 @@ bool Pattern::matches(
 			return false;
 		assertThrow(m_arguments.size() == instrAndArgs->second->size(), OptimizerException, "");
 		for (size_t i = 0; i < m_arguments.size(); ++i)
-			if (!m_arguments[i].matches(instrAndArgs->second->at(i), _dialect, _ssaValues))
+		{
+			Expression const& arg = instrAndArgs->second->at(i);
+			// If this is a direct function call instead of a variable or literal,
+			// we reject the match because side-effects could prevent us from
+			// arbitrarily modifying the code.
+			if (
+				holds_alternative<FunctionCall>(arg) ||
+				!m_arguments[i].matches(arg, _dialect, _ssaValues)
+			)
```

Comment:
If we consider the pattern `Builtins::NOT(Builtins::NOT(X))`, will this get matched any more against `not(not(1))`?  Because `arg` here would be the function call `not(1)`.

Reply:
It would not, but it would match against
```
let x := not(1)
let y := not(x)
```
which would be simplified to
```
let x := not(1)
let y := 1
```